File: backend/src/services/profileService.js
// src/services/profileService.js
// Perfil pessoal — mesmo formato para todas as personas. Devolve o utilizador,
// a empresa e um resumo adequado ao papel (comprador, fornecedor, financeiro,
// admin do sistema) + atividade recente.
const prisma = require('../config/database');

const num = (d) => Number(d ?? 0);
const RECEIVED = ['RECEBIDA_CONFORME', 'RECEBIDA_COM_DIVERGENCIA', 'CONCLUIDA'];
const itemsCount = (po) => po.items.reduce((s, i) => s + i.quantity, 0);

async function getProfile({ userId, companyId }) {
  const [user, company] = await Promise.all([
    prisma.user.findUnique({ where: { id: userId }, select: { id: true, name: true, email: true, role: true, avatarUrl: true, createdAt: true } }),
    companyId ? prisma.company.findUnique({ where: { id: companyId } }) : null,
  ]);
  const now = new Date();
  const yearStart = new Date(now.getFullYear(), 0, 1);

  // Admin do Sistema — sem empresa: estatísticas da plataforma.
  if (!company) {
    const [companies, orders, policies, users] = await Promise.all([
      prisma.company.count(),
      prisma.purchaseOrder.count(),
      prisma.supplierToKiximaPolicy.count(),
      prisma.user.count(),
    ]);
    return {
      user, company: null, kind: 'admin',
      cards: [
        { icon: 'building', tone: 'info', label: 'Empresas', value: companies, sub: 'Registadas' },
        { icon: 'orders', tone: 'success', label: 'Ordens (plataforma)', value: orders, sub: 'Total' },
        { icon: 'policy', tone: 'pending', label: 'Apólices', value: policies, sub: 'Fornecedores' },
        { icon: 'users', tone: 'info', label: 'Utilizadores', value: users, sub: 'Total' },
      ],
      recent: [],
    };
  }

  const isSupplier = company.type === 'FORNECEDOR';
  const where = isSupplier ? { supplierCompanyId: companyId } : { buyerCompanyId: companyId };
  const orders = await prisma.purchaseOrder.findMany({
    where, include: { items: true, buyerCompany: { select: { name: true } }, supplierCompany: { select: { name: true } } },
    orderBy: { updatedAt: 'desc' },
  });
  const yearOrders = orders.filter((o) => new Date(o.createdAt) >= yearStart);
  const counterpart = new Set(orders.map((o) => (isSupplier ? o.buyerCompanyId : o.supplierCompanyId)));
  const itemsMoved = orders.filter((o) => RECEIVED.includes(o.status)).reduce((s, o) => s + itemsCount(o), 0);
  const totalValue = yearOrders.reduce((s, o) => s + num(o.totalAmount), 0);

  const cards = isSupplier
    ? [
      { icon: 'orders', tone: 'info', label: 'Ordens Recebidas', value: yearOrders.length, sub: 'Este ano' },
      { icon: 'payment', tone: 'success', label: 'Valor Total Vendido', value: totalValue, money: true, sub: 'Este ano' },
      { icon: 'suppliers', tone: 'info', label: 'Clientes', value: counterpart.size, sub: 'Com transações' },
      { icon: 'reception', tone: 'pending', label: 'Itens Entregues', value: itemsMoved, sub: 'Total' },
    ]
    : [
      { icon: 'orders', tone: 'info', label: 'Ordens de Compra', value: yearOrders.length, sub: 'Este ano' },
      { icon: 'payment', tone: 'success', label: 'Valor Total Comprado', value: totalValue, money: true, sub: 'Este ano' },
      { icon: 'suppliers', tone: 'info', label: 'Fornecedores', value: counterpart.size, sub: 'Com transações' },
      { icon: 'reception', tone: 'pending', label: 'Itens Recebidos', value: itemsMoved, sub: 'Total' },
    ];

  const recent = orders.slice(0, 5).map((o) => ({
    reference: o.reference, party: isSupplier ? o.buyerCompany?.name : o.supplierCompany?.name,
    status: o.status, at: o.updatedAt,
  }));

  return { user, company, kind: isSupplier ? 'supplier' : 'buyer', cards, recent };
}

module.exports = { getProfile };

File: backend-java/src/main/java/ao/kixima/user/ProfileService.java
package ao.kixima.user;

import ao.kixima.common.error.NotFoundException;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.company.CompanyType;
import ao.kixima.company.SupplierToKiximaPolicyRepository;
import ao.kixima.po.PoStatus;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.po.PurchaseOrderItem;
import ao.kixima.po.PurchaseOrderRepository;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Espelha profileService.js — perfil pessoal, mesmo formato para todas as
 * personas: utilizador, IDENTIFICAÇÃO da empresa (nunca a ficha completa) e um
 * resumo adequado ao papel + atividade recente.
 */
@Service
public class ProfileService {

    static final Set<PoStatus> RECEIVED = Set.of(PoStatus.RECEBIDA_CONFORME, PoStatus.RECEBIDA_COM_DIVERGENCIA, PoStatus.CONCLUIDA);

    public record UserRef(String id, String name, String email, String role, String avatarUrl, Instant createdAt) {
    }

    /** Só a identificação da empresa — a ficha completa é a tela de Perfil da Empresa (Company Admin). */
    public record CompanyRef(String id, String name, String taxId, String type, String status, String contactEmail, String contactPhone,
                             String address, String city, String province, String country, String logoUrl, boolean verified) {
        public static CompanyRef de(Company c) {
            return new CompanyRef(c.getId(), c.getName(), c.getTaxId(), c.getType().name(), c.getStatus().name(), c.getContactEmail(),
                    c.getContactPhone(), c.getAddress(), c.getCity(), c.getProvince(), c.getCountry(), c.getLogoUrl(), c.isVerified());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record Card(String icon, String tone, String label, Object value, Boolean money, String sub) {
    }

    public record Recent(String reference, String party, String status, Instant at) {
    }

    /** `company` é sempre presente (null para o Admin do Sistema) — daí não excluir nulos. */
    public record Profile(UserRef user, CompanyRef company, String kind, List<Card> cards, List<Recent> recent) {
    }

    private final UserRepository userRepository;
    private final CompanyRepository companyRepository;
    private final PurchaseOrderRepository purchaseOrderRepository;
    private final SupplierToKiximaPolicyRepository supplierPolicyRepository;

    public ProfileService(UserRepository userRepository, CompanyRepository companyRepository,
                          PurchaseOrderRepository purchaseOrderRepository, SupplierToKiximaPolicyRepository supplierPolicyRepository) {
        this.userRepository = userRepository;
        this.companyRepository = companyRepository;
        this.purchaseOrderRepository = purchaseOrderRepository;
        this.supplierPolicyRepository = supplierPolicyRepository;
    }

    static int itemsCount(PurchaseOrder po) {
        return po.getItems().stream().mapToInt(PurchaseOrderItem::getQuantity).sum();
    }

    static Instant inicioDoAno() {
        return LocalDate.now(ZoneId.systemDefault()).withDayOfYear(1).atStartOfDay(ZoneId.systemDefault()).toInstant();
    }

    @Transactional(readOnly = true)
    public Profile getProfile(String userId, String companyId) {
        User u = userRepository.findById(userId).orElseThrow(() -> new NotFoundException("Utilizador"));
        UserRef user = new UserRef(u.getId(), u.getName(), u.getEmail(), u.getRole().name(), u.getAvatarUrl(), u.getCreatedAt());
        Company company = companyId == null ? null : companyRepository.findById(companyId).orElse(null);

        // Admin do Sistema — sem empresa: estatísticas da plataforma.
        if (company == null) {
            List<Card> cards = List.of(
                    new Card("building", "info", "Empresas", companyRepository.count(), null, "Registadas"),
                    new Card("orders", "success", "Ordens (plataforma)", purchaseOrderRepository.count(), null, "Total"),
                    new Card("policy", "pending", "Apólices", supplierPolicyRepository.count(), null, "Fornecedores"),
                    new Card("users", "info", "Utilizadores", userRepository.count(), null, "Total"));
            return new Profile(user, null, "admin", cards, List.of());
        }

        boolean isSupplier = company.getType() == CompanyType.FORNECEDOR;
        List<PurchaseOrder> orders = isSupplier
                ? purchaseOrderRepository.findBySupplierCompanyIdOrderByUpdatedAtDesc(companyId)
                : purchaseOrderRepository.findByBuyerCompanyIdOrderByUpdatedAtDesc(companyId);
        Instant yearStart = inicioDoAno();
        List<PurchaseOrder> yearOrders = orders.stream().filter(o -> !o.getCreatedAt().isBefore(yearStart)).toList();
        Set<String> counterpart = new HashSet<>();
        orders.forEach(o -> counterpart.add(isSupplier ? o.getBuyerCompanyId() : o.getSupplierCompanyId()));
        int itemsMoved = orders.stream().filter(o -> RECEIVED.contains(o.getStatus())).mapToInt(ProfileService::itemsCount).sum();
        BigDecimal totalValue = yearOrders.stream().map(PurchaseOrder::getTotalAmount).reduce(BigDecimal.ZERO, BigDecimal::add);

        List<Card> cards = isSupplier
                ? List.of(
                new Card("orders", "info", "Ordens Recebidas", yearOrders.size(), null, "Este ano"),
                new Card("payment", "success", "Valor Total Vendido", ao.kixima.common.Decimais.numero(totalValue), true, "Este ano"),
                new Card("suppliers", "info", "Clientes", counterpart.size(), null, "Com transações"),
                new Card("reception", "pending", "Itens Entregues", itemsMoved, null, "Total"))
                : List.of(
                new Card("orders", "info", "Ordens de Compra", yearOrders.size(), null, "Este ano"),
                new Card("payment", "success", "Valor Total Comprado", ao.kixima.common.Decimais.numero(totalValue), true, "Este ano"),
                new Card("suppliers", "info", "Fornecedores", counterpart.size(), null, "Com transações"),
                new Card("reception", "pending", "Itens Recebidos", itemsMoved, null, "Total"));

        List<Recent> recent = orders.stream().limit(5).map(o -> new Recent(o.getReference(),
                isSupplier ? nome(o.getBuyerCompany()) : nome(o.getSupplierCompany()), o.getStatus().name(), o.getUpdatedAt())).toList();
        return new Profile(user, CompanyRef.de(company), isSupplier ? "supplier" : "buyer", cards, recent);
    }

    private static String nome(Company c) {
        return c == null ? null : c.getName();
    }
}
